File: solutions/python/arrays_and_hashing/valid_sudoku.py
"""36. Valid Sudoku — https://leetcode.com/problems/valid-sudoku/

Decide whether a partially filled 9x9 board breaks any Sudoku rule.

Only the filled cells are checked, and the board need not be solvable: a board with no
contradiction among its current entries is valid. Empty cells are the string `"."`.
"""

from __future__ import annotations

from collections.abc import Iterable

from neetcode_core.registry import solution

SLUG = "valid-sudoku"

EMPTY = "."
SIZE = 9
BOX = 3


def _all_distinct(cells: Iterable[str]) -> bool:
    """True when the filled cells hold no repeat. Empty cells are ignored entirely."""
    filled = [cell for cell in cells if cell != EMPTY]
    return len(filled) == len(set(filled))


@solution(SLUG, approach="three-pass")
def valid_sudoku_three_pass(board: list[list[str]]) -> bool:
    """Rows, then columns, then boxes. Time O(1), space O(1).

    The board is a fixed 81 cells, so every approach here is O(1) and the only thing that
    varies is the constant factor. This one reads the board three times and is by far the
    easiest version to convince yourself is correct — each sweep checks exactly one rule.
    """
    if any(not _all_distinct(row) for row in board):
        return False

    for column in range(SIZE):
        if not _all_distinct(board[row][column] for row in range(SIZE)):
            return False

    for box_row in range(0, SIZE, BOX):
        for box_col in range(0, SIZE, BOX):
            cells = (
                board[box_row + r][box_col + c]
                for r in range(BOX)
                for c in range(BOX)
            )
            if not _all_distinct(cells):
                return False

    return True


@solution(SLUG, approach="single-pass")
def valid_sudoku_single_pass(board: list[list[str]]) -> bool:
    """One visit per cell, updating three trackers at once. Time O(1), space O(1).

    The box index is `(row // BOX) * BOX + column // BOX`. Worth deriving rather than
    memorising: integer-dividing a coordinate by 3 collapses it to which band of three it falls
    in, and the row band is then scaled by 3 to leave room for the column band.

    27 sets — nine rows, nine columns, nine boxes — each holding at most nine entries. That
    fixed bound is the O(1) space.
    """
    rows: list[set[str]] = [set() for _ in range(SIZE)]
    columns: list[set[str]] = [set() for _ in range(SIZE)]
    boxes: list[set[str]] = [set() for _ in range(SIZE)]

    for row, cells in enumerate(board):
        for column, cell in enumerate(cells):
            if cell == EMPTY:
                continue

            box = (row // BOX) * BOX + column // BOX
            if cell in rows[row] or cell in columns[column] or cell in boxes[box]:
                return False

            rows[row].add(cell)
            columns[column].add(cell)
            boxes[box].add(cell)

    return True

File: solutions/typescript/arrays-and-hashing/valid-sudoku.ts
/**
 * 36. Valid Sudoku — https://leetcode.com/problems/valid-sudoku/
 *
 * Decide whether a partially filled 9x9 board breaks any Sudoku rule. Only the filled cells are
 * checked, and the board need not be solvable. Empty cells are the string `"."`.
 */

import { defineProblem } from "../define-problem";

export const SLUG = "valid-sudoku";

const EMPTY = ".";
const SIZE = 9;
const BOX = 3;

/** True when the filled cells hold no repeat. Empty cells are ignored entirely. */
function allDistinct(cells: readonly string[]): boolean {
  const filled = cells.filter((cell) => cell !== EMPTY);
  return new Set(filled).size === filled.length;
}

/**
 * Rows, then columns, then boxes. Time O(1), space O(1).
 *
 * The board is a fixed 81 cells, so every approach is O(1) and only the constant factor varies.
 * This one reads the board three times and each sweep checks exactly one rule.
 */
export function validSudokuThreePass(board: readonly (readonly string[])[]): boolean {
  for (const row of board) {
    if (!allDistinct(row)) return false;
  }

  for (let column = 0; column < SIZE; column++) {
    const cells = Array.from({ length: SIZE }, (_, row) => board[row]![column]!);
    if (!allDistinct(cells)) return false;
  }

  for (let boxRow = 0; boxRow < SIZE; boxRow += BOX) {
    for (let boxCol = 0; boxCol < SIZE; boxCol += BOX) {
      const cells: string[] = [];
      for (let r = 0; r < BOX; r++) {
        for (let c = 0; c < BOX; c++) cells.push(board[boxRow + r]![boxCol + c]!);
      }
      if (!allDistinct(cells)) return false;
    }
  }

  return true;
}

/**
 * One visit per cell, updating three trackers at once. Time O(1), space O(1).
 *
 * The box index is `Math.floor(row / BOX) * BOX + Math.floor(column / BOX)`. Note the two
 * explicit `Math.floor` calls — JavaScript's `/` is floating-point division, so unlike Python's
 * `//` and PHP's `intdiv`, truncation has to be asked for. `7 / 3` is `2.333…`, and using it as
 * an array index would silently return `undefined`.
 */
export function validSudokuSinglePass(board: readonly (readonly string[])[]): boolean {
  const rows = Array.from({ length: SIZE }, () => new Set<string>());
  const columns = Array.from({ length: SIZE }, () => new Set<string>());
  const boxes = Array.from({ length: SIZE }, () => new Set<string>());

  for (let row = 0; row < board.length; row++) {
    const cells = board[row]!;
    for (let column = 0; column < cells.length; column++) {
      const cell = cells[column]!;
      if (cell === EMPTY) continue;

      const box = Math.floor(row / BOX) * BOX + Math.floor(column / BOX);
      if (rows[row]!.has(cell) || columns[column]!.has(cell) || boxes[box]!.has(cell)) {
        return false;
      }

      rows[row]!.add(cell);
      columns[column]!.add(cell);
      boxes[box]!.add(cell);
    }
  }

  return true;
}

export const validSudoku = defineProblem(SLUG, {
  "three-pass": validSudokuThreePass,
  "single-pass": validSudokuSinglePass,
});
